### ISSUE TITLE
Introduce label system for plugins

### ISSUE THREAD

**Issue by cburgdorf** [ethereum/py-evm#1545]
### What is wrong?

Today, trinity supports two different node types:

- Ethereum 1.0 full node
- Ethereum 1.0 light node

In the future I believe that we want to support at least two more node types with this code base:

- Ethereum 2.0 / Serenity beacon node
- Ethereum 2.0 / Serenity validator node
- ???

There's already PR #1536 that adds support for a `trinity-beacon` command to demonstrate one possible path to maintain an entirely different node type in this code base.

The way I see our client architecture evolving Trinity is basically layered in:

- Reusable APIs / Primitives
- Plugins
- Core (DB Access, Event Bus, Plugin Infrastructure, Logging)

We end up with a relatively lightweight core, a set of libraries and plugins that wire things together.

In this kind of architecture, not every plugin ends up being useful for every node type. Some of our plugins may be useful for every node type e.g.

- `trinity attach`
- `trinity fix-unclean-shutdown` (also, in the long term I hope we can obsolete that one)

Some of them may only be useful for an Ethereum 1 light node e.g.

- `LightPeerChainBridge`

Even today already ensure that some plugins would only run in some context e.g.

https://github.com/ethereum/py-evm/blob/8ba5ed751926ae993487aed10b3b7c65aa6a8cde/trinity/plugins/builtin/light_peer_chain_bridge/plugin.py#L43-L45

However, we not only need to keep plugins from running in the wrong context, some plugins simply should not be loaded at all so that they don't even show up in the context of `--help`.

### How can it be fixed

Introduce some form of label system for plugins. Similar how plugins advertise their name, they could advertise labels e.g.

```python
    @property
    def labels(self) -> Tuple[str,..]:
        return (PluginLabels.ETHEREUM1_FULL_NODE, PluginLabels.ETHEREUM1_LIGHT_NODE)
```

This would give nodes (even 3rd party nodes that others may build) the flexibility to simply filter the plugins that are registered at a `PluginManager` instance for specific labels.

**Comment by cburgdorf** [comment]
This is flagged as "Good First Issue" but it needs some discussion before we decide if we want to go that way. /cc @carver @pipermerriam

**Comment by pipermerriam** [comment]
I'm :-1: on "Good First Issue".

**Comment by pipermerriam** [comment]
I'd like to put this into context with the boot process.  It looks something like:

1. initial plugin setup.
2. plugins modify CLI parser
3. CLI parser parses arguments
4. convert parsed arguments to trinity config
5. either run `trinity_boot` process or other executable from CLI parser.

This issues seems to want to add a step between 1) "initial plugin setup" and 2) "plugins modify CLI parser" so that a plugin can exclude itself from the cli parser **if** it's not appropriate to be present based on the command being run.  However, this is problematic because the command isn't *knowable* until after step 3.

So it's unclear to me how we could implement this.

**Comment by carver** [comment]
> This issues seems to want to add a step between 1) "initial plugin setup" and 2) "plugins modify CLI parser" ... However, this is problematic because the command isn't knowable until after step 3.

Right, I guess the original command (eg~ `trinity` vs `trinity-beacon`) could affect which plugins are loaded, because it could use a different parser entirely. Then you don't get the light-vs-full plugin differentiation, though. At which point the benefits of the labeling seems slim. 

I guess the implication is that maybe it's just fine to decide at plugin boot time whether to run or shutdown, rather than add labeling. Any other use cases besides the CLI `--help` that labels would help with?

---

Meta: I removed Good First Issue. Let's have this be the "flesh out the idea & API" issue, and once it's solid, we can implement the labels in one or two plugins. After that's reviewed and merged, then we can add another issue that extends the API to all plugins, and that could be a GFI.

**Comment by cburgdorf** [comment]
My initial idea was to just filter what goes in here based on the labels

https://github.com/ethereum/py-evm/blob/de91bc13f47d020d26854f3d14fb9b5898845451/trinity/main.py#L380

But now that I thought about it a bit more I think I agree that it is unnecessary complex basically because for *built-in* plugins we could just hardcode the used plugins in the bootstrap code of each different node.

I think it is slightly more complicated for external plugins. Let's assume that a user wants to run an Ethereum 1 node as well as a Ethereum 2 beacon node.

Should these be two different applications installed into two different virtual directories? If yes, then the separation of external plugins comes natural as ultimately the user will just install plugin A in the installation directory of the Ethereum 1 node and plugin B in the installation directory of the beacon node.

If however, we aim to support running a beacon node and an ethereum 1 node from one installation directory, then we will need some form of management to decide which plugins are being used for which program.

So this either means:

- each plugin needs to decide which node it supports
- each plugin uses something like labels
- we use some form of config file with different sections for each supported node in which plugins can be configured per node

Anyway, my main motivation for now is to move the Ethereum 2.0 stuff forward and hence come up with a short term solution to not have incompatible plugins start with the `trinity-beacon` command and I think that doesn't need any of the more sophisticated solutions just yet.

I'll propose something in a PR that we can discuss futher. Closing here.
